File: OOP/applications/Linked lists/linked_lists.py
# -*- coding: utf-8 -*-
"""
Created on Sat Jun 29 00:30:00 2024

@author: zakar
"""

class Node : 
    def __init__(self, data):
            self.data = data
            self.next = None
            
            
class SinglyLL:
    
    def __init__(self):
        self.head= None
        self.tail = None
        self.length= 0
        
        
    def push(self,data):
        if (self.head):
            self.tail.next = Node(data)
            self.tail = self.tail.next
        else:
            self.tail = Node(data)
            self.head = self.tail
        self.length+=1
            
    def printll(self):
        nover = self.head
        while (nover):
            print(nover.data, end = " ")
            nover=nover.next
        print()
            
        
    def pop(self):
        nover = self.head
        while(nover.next.next):
            nover = nover.next
        del self.tail
        self.tail = nover
        self.tail.next = None
        self.length -=1
        
    
    
    def insert(self,ind,data):
        if 0 < ind <= self.length:
            nover = self.head
            new = Node(data)
            for i in range(ind-1):
                nover = nover.next
            c=nover.next
            nover.next=new
            new.next = c
            self.length+=1
        elif ind ==0 :
            new = Node(data)
            new.next =self.head
            self.head = new
            self.length +=1
        else:
            raise IndexError()

    

    def remove(self,ind):
        if 0 < ind <= self.length :
            nover = self.head
            for i in range(ind-1):
                nover=nover.next
            h=nover.next.next
            del nover.next
            nover.next = h
            self.length-=1
        elif ind ==0:
            h= self.head.next
            del self.head
            self.head =h
            self.length -=1
        else :
            raise IndexError()
        
        
    def __len__(self):
        return self.length


llist = SinglyLL()
for i in range(7):
    llist.push(i)
llist.printll()

llist.remove(5)
llist.printll()

print(len(llist))

File: OOP/applications/Linked lists/linked_lists.cpp
#include <iostream>
using namespace std;

class Node{
    public :
    int data;
    Node *next;
    Node(int a):data(a)
    {
        next = nullptr;
    }
};

class SinglyLinkedlist {
public :
    Node *head;
    Node *tail;
    int llsize=1;

    SinglyLinkedlist(int a){
        tail = new Node(a);
        head = tail;
    }

    void push(int a){
        Node *newnode = new Node(a);
        if (head){
            this->tail->next = newnode;
            this->tail = newnode;
        }
        else{
            this->head= newnode;
            this->tail= newnode;
        }
        this->llsize++;
    }


    void pop(){
        if (this->llsize > 1){
            Node *p = this->head;

            while(p->next != this->tail)
            {
                p=p->next;
            }
            delete this->tail;
            p->next=nullptr;
            this->tail = p;
            this->llsize--;
        }
        else if (this->llsize ==1){
                delete this->tail;
                this->head = nullptr;
                this->tail = nullptr;
                this->llsize--;
                }
        else{cout << "linkedlist is empty";}
    }

    void sizell()
    {
        cout << this->llsize <<endl;
    }

    void printll(){
    Node *p=this->head;
    if(this->llsize == 0){cout << "empty linked list"<<endl;}
    else
    {
    while(p)
            {
                cout << p->data  << " ";
                p=p->next;
            }
        cout <<endl;
    }
}
};

int main()
{
    SinglyLinkedlist kk(0);
    kk.push(1);
    kk.push(2);
    kk.push(3);
    kk.sizell();
    kk.printll();
    cout << (kk.head->next)->data;
    return 0;
}
